File: 2021_14.py
import sys;
import logging;

AOC_DAY=14
AOC_YEAR=2021

def read_formulae(logger):
    formulae = {};
    line = sys.stdin.readline().strip();
    while(line):
        pair, insertch = line.split(' -> ');
        formulae[pair] = insertch;
        line = sys.stdin.readline().strip();

    return formulae;

def calculate_polymer(formulae, initial_polymer):
    retval = initial_polymer[0]
    for pairindex in range(len(initial_polymer)-1):
        retval += formulae[initial_polymer[pairindex:pairindex + 2]] \
                + initial_polymer[pairindex + 1];
    return retval;

def count_chars(chars, polymer):
    for ch in polymer:
        chars[ch] = chars.get(ch, 0) + 1;


if('__main__' == __name__):
    import argparse;

    # deal with command line arguments
    parser = argparse.ArgumentParser(
            description = 'Advent of Code {} day {} solution'.format(
                AOC_YEAR, AOC_DAY));
    logger = logging.Logger('AOC{}d{}'.format(AOC_YEAR, AOC_DAY));
    logger.addHandler(logging.StreamHandler());
    logger.setLevel(logging.INFO);

    parser.add_argument(
        '-2', '--part2', action='store_true',
        help='part 2?', default=False);
    parser.add_argument(
        '-v', '--verbose', action='store_true',
        help='be verbose', default=False);

    args = parser.parse_args();
    if(args.verbose):
        logger.setLevel(logging.DEBUG);

    # read the initial polymer
    initial_polymer = sys.stdin.readline().strip();
    # skip a line
    sys.stdin.readline();
    # read the directions
    formulae = read_formulae(logger);

    # calculate the result for each pair
    nsteps = 10;
    # we initialize this array with the first character
    chars = { initial_polymer[0]: 1 };
    for i in range(len(initial_polymer) - 1):
        step_polymer = initial_polymer[i:i+2];
        print("processing step_polymer {}".format(step_polymer));
        for step in range(nsteps):
            step_polymer = calculate_polymer(formulae, step_polymer);
        # don't count the first character, which was counted in the last polymer
        count_chars(chars, step_polymer[1:]);

    # find max and min occurrences
    (maxch, maxoccurrences) = chars.popitem();
    (minch, minoccurrences) = (maxch, maxoccurrences);
    for (ch, occurrences) in chars.items():
        if(occurrences > maxoccurrences):
            maxch = ch;
            maxoccurrences = occurrences;
        if(occurrences < minoccurrences):
            minch = ch;
            minoccurrences = occurrences;

    logger.debug(f'After {nsteps} steps, max {maxch} with {maxoccurrences}' \
            f', min {minch} with {minoccurrences}');

    print(maxoccurrences - minoccurrences);

File: 2021_14.rb
require 'logger'
require 'scanf'

AOCDAY = 14
AOCYEAR = 2021

class Polymer
  def initialize(fh, logger)
    # initialize by reading in initial string and expansion rules
    #  initial string
    @pairs = {}
    @pairs.default = 0
    initial_string = fh.readline.strip
    (1...initial_string.length).each do |endidx|
      @pairs[initial_string[endidx - 1..endidx]] += 1
    end
    @firstletter = initial_string[0]

    #  skip blank line
    line = fh.readline

    #  expansion rules
    @rules = {}
    fh.each_line do |line|
      pair, addch = line.strip.scanf("%2s -> %c")
      @rules[pair] = [pair[0] + addch, addch + pair[1]]
    end

    @logger = logger
    @logger.debug("Rules: #{@rules}");
    @logger.debug("Initial pairs: #{@pairs}");
    @step = 0
  end

  def step()
    newpairs = {}
    newpairs.default = 0
    @pairs.each do |oldpair, n|
      # expand this pair
      newpair1, newpair2 = @rules[oldpair]
      @logger.debug("#{oldpair} -> #{newpair1}, #{newpair2} #{n}");
      # record new amounts
      newpairs[newpair1] += n
      newpairs[newpair2] += n
    end
    @pairs = newpairs
    @step += 1
    @logger.debug("Step #{@step}: #{@pairs}")
  end

  def answer
    # count of most common element - count of least common element

    # calculate count for each element. Note we only count the second letter of
    # each pair. The first letter in the string remains constant, so we
    # count that as a special case.
    elementcount = { @firstletter => 1 }
    elementcount.default = 0
    @pairs.each do |pair, newcount|
      elementcount[pair[1]] += newcount
    end
    @logger.debug("Total element counts: #{elementcount}")

    # calculate most common element - least common element
    #   one pass, rather than using ruby min/max (2 passes)
    puts elementcount
    mincount = maxcount = elementcount.first[1]
    elementcount.each do |element, ecount|
      maxcount = [ecount, maxcount].max
      mincount = [ecount, mincount].min
    end
    maxcount - mincount
  end
end


def run(inputfile, logger, **kwargs)
  polymer = nil
  File.open(inputfile) do |fh|
    polymer = Polymer.new(fh, logger)
  end

  (1..10).each do
    polymer.step
  end
  puts("Count at step 10: #{polymer.answer}")

  (11..40).each do
    polymer.step
  end
  puts("Count at step 40: #{polymer.answer}")
end


if __FILE__ == $0
  require 'optparse'

  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: $0 [options] <inputfile> [<more inputfiles>]\n" \
      "  Advent of Code #{AOCYEAR} day #{AOCDAY}"

    opts.on("-v", "--verbose", "Be verbose") do |v|
      options[:verbose] = v
    end

    opts.on("-2", "--part2", "Run part 2") do |part2|
      options[:part2] = true
    end

    opts.on("-h", "--help", "Print this help message") do
      puts opts
      exit
    end
  end.parse!

  logger = Logger.new(STDOUT);
  logger.level = (options.include? :verbose) ? Logger::DEBUG : Logger::WARN

  ARGV.each do |inputfile|
    run(inputfile, logger, **options)
  end
end
